aws-events-rule/src/main/java/software/amazon/events/rule/Translator.java

```diff
@@ -0,0 +1,795 @@
+      try {
+        eventPattern = MAPPER.writeValueAsString(model.getEventPattern());
+      } catch (final JsonProcessingException e) {
+        throw new TerminalException(e);
+      }
+    }
+
+    return putRuleRequestBuilder
+            .name(compositeId.ruleName)
+            .eventBusName(compositeId.eventBusName)
+            .description(model.getDescription())
+            .eventPattern(eventPattern)
+            .roleArn(model.getRoleArn())
+            .scheduleExpression(model.getScheduleExpression())
+            .state(model.getState())
+            .build();
+  }
+
+  /**
+   * Generates a PutTargetsRequest based on a ResourceModel.
+   * @param model A ResourceModel containing data to make a PutTargetsRequest
+   * @return A PutTargetsRequest
+   */
+  static PutTargetsRequest translateToPutTargetsRequest(final ResourceModel model) {
+    PutTargetsRequest putTargetsRequest = null;
+
+    if (model.getTargets() != null) {
+      CompositeId compositeId = new CompositeId(model);
+
+      ArrayList<Target> targets = new ArrayList<>();
+
+      for (software.amazon.events.rule.Target target : model.getTargets()) {
+        Target.Builder targetBuilder = Target.builder();
+
+        addBatchParameters(targetBuilder, target.getBatchParameters());
+        addDeadLetterConfig(targetBuilder, target.getDeadLetterConfig());
+        addEcsParameters(targetBuilder, target.getEcsParameters());
+        addHttpParameters(targetBuilder, target.getHttpParameters());
+        addInputTransformer(targetBuilder, target.getInputTransformer());
+        addKinesisParameters(targetBuilder, target.getKinesisParameters());
+        addRedshiftDataParameters(targetBuilder, target.getRedshiftDataParameters());
+        addRetryPolicy(targetBuilder, target.getRetryPolicy());
+        addRunCommandParameters(targetBuilder, target.getRunCommandParameters());
+        addSqsParameters(targetBuilder, target.getSqsParameters());
+        addSageMakerPipelineParameters(targetBuilder, target.getSageMakerPipelineParameters());
+
+        targets.add(targetBuilder
+                .arn(target.getArn())
+                .id(target.getId())
+                .input(target.getInput())
+                .inputPath(target.getInputPath())
+                .roleArn(target.getRoleArn())
+                .build()
+        );
+      }
+
+      putTargetsRequest = PutTargetsRequest.builder()
+              .eventBusName(compositeId.eventBusName)
+              .rule(compositeId.ruleName)
+              .targets(targets)
+              .build();
+    }
+
+    return putTargetsRequest;
+  }
+
+  // READ
+
+  /**
+   * Generates a DescribeRuleRequest based on a ResourceModel.
+   * @param model A ResourceModel containing data to make a DescribeRuleRequest
+   * @return A DescribeRuleRequest
+   */
+  static DescribeRuleRequest translateToDescribeRuleRequest(final ResourceModel model) {
+    CompositeId compositeId = new CompositeId(model);
```

Comment:
Because the composite id can be found either from the name and eventBusName properties or the arn property depending on what's available. During regular handler calls the compositeId is found from the name and eventBusName, but when the read Handler is called using the primaryId, the arn is provided instead.

Reply:
ohhhhhh, ok thank you! Yes then everything you're doing with the CompositeId is great the way it is